### PR TITLE
Handle copy on empty str/length zero data

### LINKED ISSUE #234: CBMC fails when empty string is passed by reference
CBMC fails with `[precondition_instance.3] memcpy destination region writeable: FAILURE` on this:
```rust
fn take_str(s: &str)) {
}

fn main() {
    take_str(&"".to_string());
}
```

The issue is that Rust seems to use a special `0x1` pointer for the empty, length-zero string. C expects `memcpy` to take valid pointers even when the length of the copy is zero. So, to match LLVM semantics, we should only codegen a memcpy when the length is nonzero.

### PR DESCRIPTION
**Pull request by avanhatt**
### Description of changes: 

Rust seems to use a special 0x1 pointer for the empty, length-zero string. C expects memcpy to take valid pointers even when the length of the copy is zero. So, to match LLVM semantics, we should only codegen a memcpy when the length is nonzero.

### Resolved issues:

Resolves #234

### Call-outs:

One of the test I adds fails not for the copy, but for some code to cast back to a ptr. This is a different issue.

### Testing:

* How is this change tested?
2 new tests, one for implicit copies and one with an intrinsic. Plus a fixme.

* Is this a refactor change?
No

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: src/test/cbmc/Strings/copy_empty_string_by_intrinsic_fixme.rs [hunk 1]
```diff
@@ -0,0 +1,42 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+// Make sure we can handle explicit copy_nonoverlapping on empty string
+
+// The copy_nonoverlapping succeeds, but the final copy back to a slice
+// fails:
+// [...copy_empty_string_by_intrinsic.assertion.2] line 1035 unreachable code: FAILURE
+// [...copy_empty_string_by_intrinsic.assertion.1] line 1037 a panicking function std::result::unwrap_failed is invoked: FAILURE
+// [...copy_string.assertion.2] line 28 assertion failed: dest_as_str.len() == l: FAILURE
```
Comment (adpaco-aws) [inline src/test/cbmc/Strings/copy_empty_string_by_intrinsic_fixme.rs:10]: Open an issue regarding these failures.

File: src/test/cbmc/Strings/copy_empty_string_by_intrinsic_fixme.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        
+        // Copy
+        let src = from_raw_parts(s.as_ptr(), l).as_ptr();
+        copy_nonoverlapping(src, dest, l);
+        
+        // THIS CHUNK causes 3 failures
+        // Back to str, check length
+        let dest_slice : &[u8] = from_raw_parts(dest, l);
+        let dest_as_str : &str= str::from_utf8(dest_slice).unwrap();
+        assert!(dest_as_str.len() == l);
```
Comment (adpaco-aws) [inline src/test/cbmc/Strings/copy_empty_string_by_intrinsic_fixme.rs:35]: Does this fail for both cases or just one? This information would be useful for debugging slice processing.

File: src/test/cbmc/Strings/copy_empty_string_by_intrinsic_fixme.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    unsafe {
+        // Unsafe buffer
+        let size: libc::size_t = size_of::<u8>();
+        let dest: *mut u8 = libc::malloc(size * l) as *mut u8;
+        
+        // Copy
+        let src = from_raw_parts(s.as_ptr(), l).as_ptr();
+        copy_nonoverlapping(src, dest, l);
+        
+        // THIS CHUNK causes 3 failures
```
Comment (adpaco-aws) [inline src/test/cbmc/Strings/copy_empty_string_by_intrinsic_fixme.rs:31]:
```suggestion
        // The chunk below causes 3 failures
```

**Review by adpaco-aws — COMMENTED** [review]
LGTM but we should keep track of the issue appearing in the intrinsic test and provide more information if possible.

File: src/test/cbmc/Strings/copy_empty_string_by_intrinsic_fixme.rs (hunk 1 above)
Comment (avanhatt) [reply src/test/cbmc/Strings/copy_empty_string_by_intrinsic_fixme.rs:11]: Added, thanks!

File: src/test/cbmc/Strings/copy_empty_string_by_intrinsic_fixme.rs (hunk 2 above)
Comment (avanhatt) [reply src/test/cbmc/Strings/copy_empty_string_by_intrinsic_fixme.rs]: Both cases, added a comment to `main` saying that, thanks.

**Review by adpaco-aws — APPROVED** [review]

**Pull request merged**